source/src/third/cuckoo/CuckooFilter.cpp

```diff
@@ -1,118 +1,62 @@
-    k *= c2;
-    h ^= k;
-    h = (h << r2) | (h >> (32 - r2));
-    h = h * m + n;
-  }
-
-  k = 0;
-  switch (key_length_in_bytes & 3) {
-    case 3: k ^= (tail[2] << 16);
-    case 2: k ^= (tail[1] << 8);
-    case 1:
-      k ^= tail[0];
-      k *= c1;
-      k = (k << r1) | (k >> (32 - r1));
-      k *= c2;
-      h ^= k;
-  }
-
-  h ^= key_length_in_bytes;
-  h ^= (h >> 16);
-  h *= 0x85ebca6b;
-  h ^= (h >> 13);
-  h *= 0xc2b2ae35;
-  h ^= (h >> 16);
-
-  return static_cast<CuckooFilter::fingerprint_type>(h);
-}
-
-/* ------------------------------------------------------------------------- */
-
-CuckooFilter::fingerprint_type CuckooFilter::scramble(fingerprint_type fingerprint) {
-	// this implementation hinges on the fact that 2**16+1 is prime,
-	// and that n-> n^17 is a bijection mod 2**16 + 1.
-	static_assert(sizeof(uint16_t) == sizeof(fingerprint_type));
-
-	uint64_t x = fingerprint; // can be uint32_t if making this 0 adjustment
-	                          // after each square-opration below.
-	if (x == 0) {
-		// lift 0 to 2**16 to get rid of silly problems involving 0.
-		x = 0x10000;
-	}
-
-	uint32_t y  = (x * x) % 0x10001; // y == x^2  mod (2**16+1)
-	y = (y * y) % 0x10001;           // y == x^4  mod (2**16+1)
-	y = (y * y) % 0x10001;           // y == x^8  mod (2**16+1)
-	y = (y * y) % 0x10001;           // y == x^16 mod (2**16+1)
-	y = (y * x) % 0x10001;           // y == x^17 mod (2**16+1)
-
-	return static_cast<fingerprint_type>(y); // intentional truncation of 0x10000 to 0.
+	return static_cast<fingerprint_type>(
+			crc16(
+					static_cast<const uint8_t*>(key),
+					key_length_in_bytes,
+					nullptr));
 }
 
 /* ------------------------------------------------------------------------- */
 
 CuckooFilter::ExtendedFingerprint CuckooFilter::getExtendedFingerprint(
-		fingerprint_type fingerprint) {
-	fingerprint_type hashed_finger = scramble(fingerprint);
+		fingerprint_type finger, index_type bucket_index) {
+
 	return ExtendedFingerprint {
-		.fingerprint = fingerprint,
-		.bucketA = static_cast<index_type>(hashed_finger % bucket_count),
-		.bucketB = static_cast<index_type>((hashed_finger ^ fingerprint) % bucket_count)
+		.fingerprint = finger,
+		.bucketA = static_cast<index_type>(bucket_index),
+		.bucketB = static_cast<index_type>((bucket_index ^ finger) % bucket_count)
 	};
 }
 
-/* ------------------------------------------------------------------------- */
+
 
 CuckooFilter::ExtendedFingerprint CuckooFilter::getExtendedFingerprint(
 		key_type key, size_t key_length_in_bytes) {
-	return getExtendedFingerprint(murmurhash(key, key_length_in_bytes));
+
+	fingerprint_type finger = hash(key, key_length_in_bytes);
+	fingerprint_type hashed_finger = hash(&finger, sizeof(finger));
+
+	return ExtendedFingerprint {
+		.fingerprint = finger,
+		.bucketA = static_cast<index_type>(hashed_finger % bucket_count),
+		.bucketB = static_cast<index_type>((hashed_finger ^ finger) % bucket_count)
```

Comment:
Since when does this work? :o

Reply:
These are called designated initializers. They will be part of c++20, but are already accepted by gnu gcc/g++ as a language extension :-D https://en.cppreference.com/w/cpp/language/aggregate_initialization

Reply:
Ah nice, didn't know we can already use them, awesome, that was a feature i was waiting for!